### PR TITLE
Allow custom options

### LINKED ISSUE #343: Allow option http-keep-alive and TLS backends
Hi,

I looked at the [the haproxy template](https://github.com/appscode/voyager/blob/3243acd62e9c8ad7806b5d6d411c67de1df41d59/pkg/ingress/template/template.go) and found some configurations that I would like to change which is currently not possible:

- support for http-keep-alive instead of http-server-close
- support for HTTPs backends like `server srv01 10.20.30.40:443 weight 1 maxconn 100 check ssl verify none`

Is there a particular reason why you chose http-server-close over http-keep-alive?

### PR DESCRIPTION
**Pull request by sadlil**
Fix #343 

Supports all valid options for defaults section of HAProxy config from the list from here
https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#4.2-option%20abortonclose

Expects a json encoded map
ie: "ingress.alpha.appscode.com/default-option": {"http-keep-alive": "true", "dontlognull": "true", "clitcpka": "false"}
This will be appended in the defaults section of HAProxy as
```
   option http-keep-alive
   option dontlognull
   no option clitcpka
```

It is upto user to provide valid configurations. Providing invalid options will not run HAProxy. The log will display something like
```
daemon.err: Aug 18 06:14:41 reloader:  [ALERT] 229/061441 (37) : parsing [/etc/haproxy/haproxy.cfg:21]: negation/default is not supported for option 'invalid-options-false'.
daemon.err: Aug 18 06:14:41 reloader: [ALERT] 229/061441 (37) : parsing [/etc/haproxy/haproxy.cfg:22] : unknown option 'invalid-options-true'.
daemon.err: Aug 18 06:14:41 reloader: [ALERT] 229/061441 (37) : Error(s) found in configuration file : /etc/haproxy/haproxy.cfg
daemon.err: Aug 18 06:14:41 reloader: [ALERT] 229/061441 (37) : Fatal errors found in configuration.
daemon.err: Aug 18 06:14:41 reloader: I0818 06:14:41.308405      18 mount.go:140] failed to run cmd

```

### CODE REVIEW

File: api/annotations.go [hunk 1]
```diff
@@ -270,6 +285,23 @@ func (r Ingress) Timeouts() map[string]string {
 	return ans
 }
 
+func (r Ingress) Options() map[string]bool {
```
Comment (tamalsaha) [inline api/annotations.go]: `HAProxyOptions()`

File: pkg/haproxy/template.go [hunk 2]
```diff
@@ -53,11 +53,10 @@ global
 defaults
 	log global
 
-	option http-server-close
-
-	# Disable logging of null connections (haproxy connections like checks).
-	# This avoids excessive logs from haproxy internals.
-	option dontlognull
+	# Defaults Options provided by user
```
Comment (tamalsaha) [inline pkg/haproxy/template.go]: `# Defaults Options provided by user` these are not always provided by the user. Remove this comment. Instead add a link to HAProxy docs.

**Comment by greg-jaunt** [comment]
@sadlil the example above and in the comment in the source code is incorrect.  You must enclose the options in single quotes, like so:

> ingress.appscode.com/default-option": '{"http-keep-alive": "true", "dontlognull": "true", "clitcpka": "false"}'

**Comment by greg-jaunt** [comment]
After more testing, I find this implementation very limiting.  It only allows settings that start with the word "option" and it seems to be somewhat arbitrary which keywords HAProxy has chosen to use "option" with.  This now allows me to specify "option http-keep-alive" but, for example, does not allow me to specify "balance leastconn".

**Comment by tamalsaha** [comment]
We have been talking about this internally. There is also a similar request in #439 . The general issue is that HAProxy has too many options and converting them all into annotations seems the wrong way to do it. One way we talked were just let users specify the haproxy.conf. That will allow anyone to do any customization they see fit. The question obviously is what will that look like?
 
One idea I have is allow users to provide the HAProxy template via a ConfigMap. We currently have a built-in template https://github.com/appscode/voyager/blob/master/pkg/haproxy/template.go#L31 . The user provided template must have the parameters that voyager operator uses to inject actual pod ips etc. But the any option, etc can be left to users to modify to their heart's content.

What do folks guys think about that?

cc: @julianvmodesto

**Comment by greg-jaunt** [comment]
If you wanted to use ConfigMaps for the config, how would you deal with the backend section, which can be dynamic when voyager pods change nodes?  Perhaps you would need a ConfigMap per section (defaults, backend, frontend).

FWIW, I was able to work around this for "balance leastconn" by adding it as a backendRule: in the ingress.

**Comment by greg-jaunt** [comment]
I suppose, alternatively, Voyager could start with the user-provided ConfigMap config and modify it either smartly somehow or based on templated variables that the user would have to add to their ConfigMap.

**Comment by tamalsaha** [comment]
I'm afraid that  this is one of those situations where there are no simple answer.

I think the way we do it we provide the templates as separate files. Each variable in https://github.com/appscode/voyager/blob/master/pkg/haproxy/template.go#L41 becomes a separate template file. The root template loads the individual pieces. All these `options` are the root template section, if you want to modify the global / defaults section. 

So, users can replace the root template to inject their custom options and keep the other templates as-is. Voyager should be able to inject the dynamic parts easily with that.

**Comment by tamalsaha** [comment]
```
# HAProxy configuration generated by https://github.com/appscode/voyager
# DO NOT EDIT!
global
	daemon
	stats socket /tmp/haproxy
	server-state-file global
	server-state-base /var/state/haproxy/
	maxconn 4000
	# log using a syslog socket
	log /dev/log local0 info
	log /dev/log local0 notice
	tune.ssl.default-dh-param 2048
	ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!3DES:!MD5:!PSK
defaults
	log global
	# https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#4.2-option%20abortonclose
	# https://github.com/appscode/voyager/pull/403
	{{- range $k, $v := .OptionsDefaults }}
	{{ if not $v }}no {{ end }}option {{ $k -}}
	{{ end }}
	# Timeout values
	{{- range $k, $v := .TimeoutDefaults }}
	timeout {{ $k }} {{ $v -}}
	{{ end }}
	# default traffic mode is http
	# mode is overwritten in case of tcp services
	mode http
{{- range $resolver := .DNSResolvers }}
{{ template "dns-resolver" $resolver -}}
{{ end }}
{{- if .Stats }}
{{ template "stats" .Stats -}}
{{ end -}}
{{- range $svc := .HTTPService }}
{{- template "http-frontend" $svc  }}
{{ template "http-backend" $svc  }}
{{ end -}}
{{- range $svc := .TCPService }}
{{ template "tcp-frontend" $svc }}
{{ template "tcp-backend" $svc }}
{{ end -}}
{{- if and (not .HTTPService) .DefaultBackend }}
{{ template "default-frontend" .SharedInfo }}
{{ end }}
{{- if .DefaultBackend }}
{{ template "default-backend" .SharedInfo }}
{{ end }}
`))
```

**Comment by tamalsaha** [comment]
I have got a pr #454 for custom templates. Voyager will by default come with templates here:
https://github.com/appscode/voyager/tree/tpl/hack/docker/voyager/templates

If you want custom template, you mount that as a configmap and pass it as a flag (glob pattern) to voyager operator (--custom-templates).

**Comment by tamalsaha** [comment]
I have added docs on how to use custom templates. https://github.com/appscode/voyager/blob/master/docs/user-guide/ingress/custom-templates.md

You can try this with `appscode/voyager:3.2.0-rc.3` .

**Pull request merged**